test/e2e/azure_securitygroups.go

```diff
@@ -0,0 +1,175 @@
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AzureSecurityGroupsSpecInput is the input for AzureSecurityGroupsSpec.
+type AzureSecurityGroupsSpecInput struct {
+	BootstrapClusterProxy framework.ClusterProxy
+	Namespace             *corev1.Namespace
+	ClusterName           string
+	Cluster               *clusterv1.Cluster
+	WaitForUpdate         []interface{}
+}
+
+func AzureSecurityGroupsSpec(ctx context.Context, inputGetter func() AzureSecurityGroupsSpecInput) {
+	var (
+		specName         = "azure-vmextensions"
+		testSecurityRule = infrav1.SecurityRule{
+			Name:             "test-security-rule",
+			Description:      "test-security-rule",
+			Protocol:         "Tcp",
+			Direction:        "Outbound",
+			Priority:         100,
+			SourcePorts:      pointer.String("*"),
+			DestinationPorts: pointer.String("80"),
+			Source:           pointer.String("*"),
+			Destination:      pointer.String("*"),
+		}
+		testSecurityRule2 = infrav1.SecurityRule{
+			Name:             "test-security-rule-2",
+			Description:      "test-security-rule-2",
+			Protocol:         "Tcp",
+			Direction:        "Inbound",
+			Priority:         110,
+			SourcePorts:      pointer.String("*"),
+			DestinationPorts: pointer.String("80"),
+			Source:           pointer.String("*"),
+			Destination:      pointer.String("*"),
+		}
+		input AzureSecurityGroupsSpecInput
+	)
+
+	Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+
+	input = inputGetter()
+	Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+	Expect(input.Namespace).ToNot(BeNil(), "Invalid argument. input.Namespace can't be nil when calling %s spec", specName)
+	Expect(input.ClusterName).ToNot(BeEmpty(), "Invalid argument. input.ClusterName can't be empty when calling %s spec", specName)
+
+	By("creating a Kubernetes client to the workload cluster")
+	workloadClusterProxy := input.BootstrapClusterProxy.GetWorkloadCluster(ctx, input.Namespace.Name, input.ClusterName)
+	Expect(workloadClusterProxy).NotTo(BeNil())
+	mgmtClient := bootstrapClusterProxy.GetClient()
+	Expect(mgmtClient).NotTo(BeNil())
+
+	azureCluster := &infrav1.AzureCluster{}
+	err := mgmtClient.Get(ctx, client.ObjectKey{
+		Namespace: input.Cluster.Spec.InfrastructureRef.Namespace,
+		Name:      input.Cluster.Spec.InfrastructureRef.Name,
+	}, azureCluster)
+	Expect(err).NotTo(HaveOccurred())
+
+	var expectedSubnets infrav1.Subnets
+	checkSubnets := func(g Gomega) {
+		azureCluster := &infrav1.AzureCluster{}
+		err = mgmtClient.Get(ctx, client.ObjectKey{
+			Namespace: input.Cluster.Spec.InfrastructureRef.Namespace,
+			Name:      input.Cluster.Spec.InfrastructureRef.Name,
+		}, azureCluster)
+		Expect(err).NotTo(HaveOccurred())
+		actualSubnets := azureCluster.Spec.NetworkSpec.Subnets
```

Comment:
I think we'll want to derive our view of "actual subnets" from the Azure APIs directly since I think reading from the AzureCluster wouldn't reveal the bug this PR fixes.